### ISSUE TITLE
Custom SameDiff ops missing from graph after upgrade to M2

### ISSUE THREAD

**Issue by partarstu** [deeplearning4j/deeplearning4j#9670]
#### Issue Description

- expected behavior - the same as for the previous releases: Custom Ops, if invoked in the graph explicitly, should remain intact.
- encountered behavior - Custom Op is ignored and replaced with the standard one having the same SameDiff name

#### Version Information

* Deeplearning4j version - M2
* Platform information (OS, etc) - Windows 10/ Linux(Buster 10)
* CUDA version, if used - N/A
* NVIDIA driver version, if in use - N/A

#### Additional Information

I’m using a custom SameDiff `Gather Op` which extends a standard (native) one by overriding the `doDiff() `method. The reason to using it this way is described [here](https://github.com/eclipse/deeplearning4j/issues/9669). Because the Op implementation itself remains the same and only the back-prop is changed, nothing else is overridden and the Op Name remains the same - "gather". 

In the code the Custom Op is added to the graph in the following way : 
`new EnhancedGatherOp(sd, lookupTable, indices, dimension).outputVariable();`

It worked all the time with different DL4J versions but after upgrading to **M2** this Op is being ignored and automatically replaced by the standard `Gather Op`. After debugging the code I noticed that it happens because of always cloning the Ops (not sure why it happens all the time, even when there’s no importing etc.) in `org/nd4j/autodiff/samediff/SameDiff.java:640`, inside `org.nd4j.autodiff.samediff.serde.FlatBuffersMapper#cloneViaSerialize(...).` The logic uses a map having the hash (which is basically a unique Native Op identifier) and a corresponding Op instance. Because my custom implementation has nothing to do with this logic it's being ignored. 

The fastest way to workaround this issue was modifying `ImportClassMapping.getOpNameMapping()` by replacing the standard op with the custom one do the trick. It needs to be done before gradient functions initialization or before loading the saved SameDiff file - otherwise the mentioned above _<hash, OpInstance>_ map will contain the mapping to the standard (native) gather op. It's quite a nasty way to register a custom Op so there should probably be something quite better. Also there's a question if cloning Ops all the time even if nothing had been imported makes sense.

**Comment by agibsonccc** [comment]
@partarstu can you do a reproducer for this one as well? Thanks!

**Comment by partarstu** [comment]
@agibsonccc the custom implementation is [here](https://gist.github.com/partarstu/776b0cf91fb7d68881c71c1becabe99b). I'll add this link to another issue as well.

The invocation (snippet of my graph) happens this way:

```
var sequenceLengthSd = sd.constant(256);
var hiddenSizeSd= sd.constant(768);
var tokenEmbeddingsMatrix = sd.var("tokenEmbeddingsMatrix ", new XavierInitScheme('c', 30522, 768), FLOAT, 30522, 768);
var inputTokenVocabIndices = sd.var("inputTokenVocabIndices  ", new XavierInitScheme('c', 512, 256), INT32, 512, 256);
var batchTokenEmbeddings = new CustomGatherOp(sd, tokenEmbeddingsMatrix, inputTokenVocabIndices, 0).outputVariable();
var batchSize = batchTokenEmbeddings.shape().get(SDIndex.point(0)).castTo("batchSize", INT32);
var flatBatchSizeSd = batchSize.mul(sequenceLengthSd);
var hiddenLayerInputShape = sd.stack("hiddenLayerInputShape", 0, flatBatchSizeSd, hiddenSizeSd);
var layerInput = batchTokenEmbeddings.reshape(hiddenLayerInputShape);
```

If you run it on M1, you'll se the log **_"Using custom Gather Op"_**. On **M2** you won't see this, because the custom Op is being ignored and the `doDiff()` of the native **Gather Op** is being called instead. 

If before initializing Grad Functions (even at the end of the graph building) the following code is added, it works:
 ```
var declaredConstructor = customOpClass.getDeclaredConstructor();
 declaredConstructor.setAccessible(true);
 var opInstance = declaredConstructor.newInstance();
 ImportClassMapping.getOpNameMapping().computeIfPresent(opInstance.opName(), (name, differentialFunction) -> opInstance);
```

Also a small comment - the snippet of code above works for Java Reflection even if the Op has a private no-arg constructor. I didn't test it with M1, but with M2 if my custom Op has a private constructor - the already mentioned cloning mechanism (`FlatBuffersMapper#cloneViaSerialize(...)`) fails, because behind the scenes it uses `Class.newInstance()` which fails if the constructor is private. That's why all Op classes in M2 must have public no-arg constructors, which is not so good for encapsulation of internal reflection-related logic.

**Comment by agibsonccc** [comment]
@partarstu actually looking at this issue again (sorry just got the gather fix done which ended up in some new features as well) I'm actually ok with ops being like that. It's not a problem for us in practice and ops themselves are normally used for configuration. I'm going to close this as a won't fix. In samediff tself, we have several hooks in DyamicCustomOp such as configureFromArguments and configureWithSamediff to allow for serialization and loading from a default constructor.
